File: src/main/java/ac/kr/deu/connect/luck/food_truck/service/FoodTruckReviewService.java
package ac.kr.deu.connect.luck.food_truck.service;

import ac.kr.deu.connect.luck.food_truck.FoodTruckMapper;
import ac.kr.deu.connect.luck.food_truck.dto.FoodTruckReviewRequest;
import ac.kr.deu.connect.luck.food_truck.dto.FoodTruckReviewResponse;
import ac.kr.deu.connect.luck.food_truck.entity.FoodTruck;
import ac.kr.deu.connect.luck.food_truck.entity.FoodTruckReview;
import ac.kr.deu.connect.luck.food_truck.repository.FoodTruckRepository;
import ac.kr.deu.connect.luck.food_truck.repository.FoodTruckReviewRepository;
import ac.kr.deu.connect.luck.image.ImageUploader;
import ac.kr.deu.connect.luck.user.User;
import ac.kr.deu.connect.luck.user.UserRepository;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class FoodTruckReviewService {

    private final FoodTruckRepository foodTruckRepository;
    private final FoodTruckReviewRepository foodTruckReviewRepository;
    private final UserRepository userRepository;
    private final ImageUploader imageUploader;
    private final FoodTruckMapper foodTruckMapper;

    /**
     * 푸드트럭 리뷰 등록
     *
     * @param foodTruckId            푸드트럭 ID
     * @param foodTruckReviewRequest 푸드트럭 리뷰 요청
     * @param userEmail              사용자 이메일
     */
    public FoodTruckReviewResponse saveFoodTruckReview(Long foodTruckId, FoodTruckReviewRequest foodTruckReviewRequest, String userEmail) {
        FoodTruck foodTruck = FoodTruck.builder().id(foodTruckId).build();
        User author = userRepository.findByEmail(userEmail).orElseThrow();

        FoodTruckReview review = foodTruckMapper.toFoodTruckReview(foodTruckReviewRequest);

        review.setFoodTruck(foodTruck);
        review.setAuthor(author);

        if (foodTruckReviewRequest.getImage() != null) {
            String imageUrl = imageUploader.uploadImage(foodTruckReviewRequest.getImage()).getData().getUrl();
            review.setImageUrl(imageUrl);
        }

        return foodTruckMapper.toFoodTruckReviewResponse(foodTruckReviewRepository.save(review));
    }

    /**
     * 푸드트럭 리뷰 수정
     *
     * @param foodTruckId            푸드트럭 ID
     * @param foodTruckReviewRequest 푸드트럭 리뷰 요청
     * @param userEmail              사용자 이메일
     * @param reviewID               리뷰 ID
     * @return 수정된 리뷰
     */
    public FoodTruckReviewResponse saveFoodTruckReview(Long foodTruckId, FoodTruckReviewRequest foodTruckReviewRequest, String userEmail, Long reviewID) {
        FoodTruckReview review = foodTruckReviewRepository.findById(reviewID).orElseThrow();

        review.setRating(foodTruckReviewRequest.getRating());
        review.setContent(foodTruckReviewRequest.getContent());

        if (foodTruckReviewRequest.getImage() != null) {
            String imageUrl = imageUploader.uploadImage(foodTruckReviewRequest.getImage()).getData().getUrl();
            review.setImageUrl(imageUrl);
        }

        return foodTruckMapper.toFoodTruckReviewResponse(foodTruckReviewRepository.save(review));
    }

    /**
     * 푸드트럭 리뷰 삭제
     *
     * @param reviewId 리뷰 ID
     */
    @Transactional
    public void deleteFoodTruckReview(Long reviewId) {
        foodTruckReviewRepository.deleteById(reviewId);
    }

    /**
     * 푸드트럭 리뷰 답글 등록
     *
     * @param reviewId 리뷰 ID
     * @param content  답글 내용
     * @return 답글이 달린 리뷰
     */
    public FoodTruckReviewResponse addReplyToReview(Long reviewId, String content) {
        FoodTruckReview review = foodTruckReviewRepository.findById(reviewId).orElseThrow();
        review.setReply(content);
        return foodTruckMapper.toFoodTruckReviewResponse(foodTruckReviewRepository.save(review));
    }

    public FoodTruckReview getFoodTruckReview(Long reviewId) {
        return foodTruckReviewRepository.findById(reviewId).get();
    }
}

File: src/main/java/ac/kr/deu/connect/luck/foodtruck/service/FoodTruckReviewService.kt
package ac.kr.deu.connect.luck.foodtruck.service

import ac.kr.deu.connect.luck.common.storage.StorageService
import ac.kr.deu.connect.luck.foodtruck.controller.request.FoodTruckReviewCreateForm
import ac.kr.deu.connect.luck.foodtruck.entity.FoodTruckReview
import ac.kr.deu.connect.luck.foodtruck.repository.FoodTruckRepository
import ac.kr.deu.connect.luck.foodtruck.repository.FoodTruckReviewRepository
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class FoodTruckReviewService(
    private val storageService: StorageService,
    private val foodTruckRepository: FoodTruckRepository,
    private val foodTruckReviewRepository: FoodTruckReviewRepository,
) {
    /**
     * 푸드트럭 리뷰를 작성합니다.
     * @param foodTruckId 푸드트럭 식별자
     * @param userId 사용자 식별자
     * @param foodTruckReviewRequest 푸드트럭 리뷰 요청 폼
     * @return 작성된 리뷰
     */
    @Transactional
    fun createReview(
        foodTruckId: Long,
        userId: Long,
        foodTruckReviewRequest: FoodTruckReviewCreateForm,
    ): FoodTruckReview {
        return foodTruckReviewRepository.save(
            FoodTruckReview.create(
                foodTruckId = foodTruckId,
                authorId = userId,
                content = foodTruckReviewRequest.content,
                images = foodTruckReviewRequest.images?.map { storageService.save(it) },
                rating = foodTruckReviewRequest.rating,
            )
        )
    }

    /**
     * 푸드트럭 리뷰를 수정합니다.
     * @param reviewId 리뷰 식별자
     * @param userId 사용자 식별자
     * @param foodTruckReviewRequest 푸드트럭 리뷰 요청 폼
     * @return 수정된 리뷰
     */
    @Transactional
    fun updateReview(
        reviewId: Long,
        userId: Long,
        foodTruckReviewRequest: FoodTruckReviewCreateForm,
    ): FoodTruckReview {
        val review = foodTruckReviewRepository.findByIdOrNull(reviewId) ?: throw IllegalArgumentException("Review not found")
        if (review.authorId != userId) {
            throw IllegalArgumentException("User is not the author of the review")
        }

        review.content = foodTruckReviewRequest.content
        review.images = foodTruckReviewRequest.images?.map { storageService.save(it) }
        review.rating = foodTruckReviewRequest.rating

        return review
    }

    /**
     * 푸드트럭 리뷰 답글 작성합니다.
     * @param reviewId 리뷰 식별자
     * @param userId 사용자 식별자(푸드트럭 매니저)
     * @param reply 답글 내용
     * @return 리뷰
     */
    @Transactional
    fun createReply(
        reviewId: Long,
        userId: Long,
        reply: String,
    ): FoodTruckReview {
        val review = foodTruckReviewRepository.findByIdOrNull(reviewId)
            ?: throw IllegalArgumentException("Review not found")

        review.reply = reply
        return review
    }


    /**
     * 푸드트럭 리뷰를 삭제합니다.
     * @param reviewId 리뷰 식별자
     * @param userId 사용자 식별자
     */
    @Transactional
    fun deleteReview(
        reviewId: Long,
        userId: Long,
    ) {
        val review = foodTruckReviewRepository.findByIdOrNull(reviewId) ?: throw IllegalArgumentException("Review not found")
        if (review.authorId != userId) {
            throw IllegalArgumentException("User is not the author of the review")
        }

        foodTruckReviewRepository.delete(review)
    }

    /**
     * 푸드트럭 리뷰를 조회합니다. (푸드트럭 리뷰 조회)
     * @param foodTruckId 푸드트럭 식별자
     * @return 푸드트럭 리뷰 목록
     */
    @Transactional(readOnly = true)
    fun getReviewsByFoodTruckId(
        foodTruckId: Long,
        pageable: Pageable,
    ): Page<FoodTruckReview> {
        return foodTruckReviewRepository.findAllByFoodTruckId(foodTruckId, pageable)
    }

    /**
     * 푸드트럭 리뷰를 조회합니다. (작성자 리뷰 조회)
     * @param userId 사용자 식별자
     * @return 작성한 리뷰 목록
     */
    @Transactional(readOnly = true)
    fun getReviewsByUserId(
        userId: Long,
        pageable: Pageable,
    ): Page<FoodTruckReview> {
        return foodTruckReviewRepository.findAllByAuthorId(userId, pageable)
    }
}
